mltensor-explainer.md

```diff
@@ -0,0 +1,421 @@
+  tensorizedGpuBufferAfterInference.destroy();
+
+  // Call this method for each frame.
+  video.requestVideoFrameCallback(applyEffectToFrame);
+}
+```
+
+## Design Discussion
+
+### Timelines
+
+WebNN uses a programming model similar to [WebGPU's](https://www.w3.org/TR/webgpu/#programming-model), in that compute tasks are posted to a timeline - which I've referred to as an "ML context timeline" throughout this document - separate from the content timeline (i.e. "script"). See [the WebGPU documentation of timelines](https://gpuweb.github.io/gpuweb/#programming-model-timelines) for more details.
+
+Specifying WebNN timelines is tracked in [#529](https://github.com/webmachinelearning/webnn/issues/529).
+
+### Device Affinity and Relationship to a `GPUDevice`
+
+The user agent decides where the memory backing an `MLTensor` is allocated. The WebNN API allows the developer to provide hints - primarily via `MLTensorUsageFlags` - but these are not binding.
+
+For example [an `MLContext` may be created with a `GPUDevice`](https://www.w3.org/TR/webnn/#dom-ml-createcontext-gpudevice), and creating an `MLTensor` from this context with the `MLTensorUsage.WEBGPU_INTEROP` flag expresses a clear intention to share the tensor with the given `GPUDevice`. However, there is no guarantee that sharing this tensor with WebGPU will be zero-copy.
+
+The `MLTensorUsage.READ_FROM` and `MLTensorUsage.WRITE_TO` flags likewise are hints to the user agent indicating that the underlying data will be read and written to, respectively, by script.
+
+### Importing an `MLTensor` to WebGPU
+
+Any `MLTensor` created with the `MLTensorUsage.WEBGPU_INTEROP` flag may be imported into any `GPUDevice`, though cross-device buffer sharing may require expensive data copies. Sharing the tensor requires coordinating between the respective WebNN and WebGPU timelines. Below is an example of how this handoff might work:
+
+- Two fences are created:
+  1. a "start access" fence which is to be signaled by WebNN and waited on by WebGPU. A data copy may be required alongside the signaling of this fence
+  2. an "end access" fence which is to be signaled by WebGPU and waited on by WebNN. A data copy may be required alongside the signaling of this fence
+- The `GPUDevice` enqueues a command to its `GPUQueue` to wait for the "start access" fence to be signaled
+- WebNN will signal the "start access" fence after the completion of all currently-enqueued operations that use the `MLTensor` which is to be imported (this is very similar to how [`GPUBuffer.mapAsync()`](https://www.w3.org/TR/webgpu/#dom-gpubuffer-mapasync) works)
+- Until the "end access" fence is signaled:
+  - The `GPUDevice` has exclusive, read/write access to the imported buffer
+  - All WebNN work involving the imported `MLTensor` is blocked
+- When the `GPUBuffer` is destroyed, the "end access" fence is signaled and the `MLTensor` may be used again by WebNN
+
+### `compute()` vs. `dispatch()`
```

Comment:
> compute() has an advantage when the tensor data starts from the CPU then ends on the CPU

Isn't this always the case? Data starts on the CPU to be passed to `writeBuffer()`, then ends on the CPU when read back via `readBuffer()`

> there is a fixed overhead per input/output which compute() can combine

What exactly is being combined? My understanding is that `compute()` has to internally:
- post a task to upload input data to input buffers
- post a task to execute the graph
- post a task to download output data to output buffers

Using `dispatch()` this maps to explicit calls to `writeBuffer()`, `dispatch()`, and `readBuffer()`. In both cases, we're posting the same number of tasks, no? Are you suggesting the delay in _posting tasks_ (since we're running more script in the latter case) is reliably longer than the tasks themselves, so the ML context may idle between tasks? Or something else?

Reply:
Using compute(), we can create a single task to perform createBuffer() for all I/O buffers and then use another single task to readBuffer() them back. In contrast, Dispatch() requires each I/O buffer operation to be its own separate task. This approach is better when I/O buffers are on-device, as tasks can be executed in parallel. However, if the I/O buffers never leave the CPU or are too small, there is little or nothing to parallelize.

Reply:
Ah I see. Added a note about this

Reply:
LGTM.